### PR TITLE
Separate out model representation of MenuItems

### LINKED ISSUE #1002: ContextMenu should derive Menu, not contain Menu

### PR DESCRIPTION
**Pull request by smoogipooo**
Closes #1002.

I'm sorry for the largeness of this PR - everything basically trickles down. We're going to go through this IRL.

Summary TBD.

### CODE REVIEW

**Comment by peppy** [comment]
we went through this IRL, but would be good to get further opinions. Still unsure about the generic `TItem`.

Also note that Dropdowns are not yet completely addressed and have only been modified to maintain existing functionality.

cc/ @Tom94 @default0

**Comment by smoogipooo** [comment]
Alright a few days late to the party, but let me give a quick rundown over what's changed here. The goal of this project was to get `Menu` to use something closer to an MVVM pattern. There are three primary concepts to my approach:  
1. Split out models from the views/view-models.
    * MenuItem is now a simple data structure representing the model. I envisioned this to be implemented with Bindables, and currently they are, however working with `.ValueChanged` functions is a little bit ugly right now so those aren't implemented yet.
    * These contain only what's necessary for the user to put data into a drawable representing a MenuItem.
2. The views/view-models are structured as protected or private nested classes within each other, with virtual methods in the parenting containers to create the "next level down".
    * In this way, `Menu` contains and provides a method to create a `DrawableMenuItem`.
    * `Dropdown` provides a method to create a `DropdownMenu`, which provides a method to create a `DrawableDropdownMenuItem`.
    * etc...
3. Encapsulate all drawing logic as much as possible - as much as `CompositeDrawable` allows us to, and provide only methods to "add" `MenuItem` models to the `Menu`.
    * The only way one can access the contained `DrawableMenuItem`s now is by overriding `Menu` and accessing the protected `Children`, but even then it is a get-only property. I believe this is adequate enough access.

I've also taken the opportunity to merge `ContextMenu` and `Menu` in this PR as much of the logic was shared. There are a few less-than-niceties remaining in `Dropdown` and `TabControl`, as @peppy mentioned, but I wanted to keep the changes as minor as possible while achieving the intended effects. Keep in mind that both of those still need to be modified to use the same/similar models.

File: osu.Framework/Graphics/UserInterface/BasicDropdown.cs [hunk 1]
```diff
@@ -56,7 +56,12 @@ public DrawableBasicDropdownMenuItem(DropdownMenuItem<T> item)
                     Foreground.Padding = new MarginPadding(2);
                 }
 
-                protected override Drawable CreateContent() => new SpriteText { Text = Item.Text };
+                protected override Drawable CreateContent()
```
[the remark this answers is not in the document]
Comment (smoogipooo) [reply osu.Framework/Graphics/UserInterface/BasicDropdown.cs]: In every single case or in the base method?

[the remark this answers is not in the document]
Comment (peppy) [reply osu.Framework/Graphics/UserInterface/BasicDropdown.cs]: base is fine

File: osu.Framework/Graphics/Sprites/IHasText.cs [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+// Licensed under the MIT Licence - https://raw.githubusercontent.com/ppy/osu-framework/master/LICENCE
+
+namespace osu.Framework.Graphics.Sprites
+{
+    /// <summary>
+    /// Interface for <see cref="IDrawable"/> components that support reading and writing text.
+    /// </summary>
+    public interface IHasText : IDrawable
+    {
+        string Text { get; set; }
```
Comment (peppy) [inline osu.Framework/Graphics/Sprites/IHasText.cs]: if we make this `set` only then `TextFlow` could also implement it (might need to be called `IHasSettableText` or something).

May be worth considering.

**Pull request merged**